iroha_crypto/src/lib.rs

```diff
@@ -0,0 +1,398 @@
+//! This module contains structures and implementations related to the c{ digest_function: (), payload: ()}digest_function: (), payload: ()}ptographic parts of the
+//! Iroha.
+
+pub mod multihash;
+
+use multihash::Multihash;
+use parity_scale_codec::{Decode, Encode};
+use serde::{de::Error as SerdeError, Deserialize};
+use std::{
+    collections::BTreeMap,
+    convert::{TryFrom, TryInto},
+    fmt::{self, Debug, Display, Formatter},
+};
+use ursa::{
+    blake2::{
+        digest::{Input, VariableOutput},
+        VarBlake2b,
+    },
+    keys::{
+        KeyGenOption as UrsaKeyGenOption, PrivateKey as UrsaPrivateKey, PublicKey as UrsaPublicKey,
+    },
+    signatures::{ed25519::Ed25519Sha512, SignatureScheme, Signer},
+};
+
+pub const SIGNATURE_LENGTH: usize = 64;
+pub const HASH_LENGTH: usize = 32;
+pub const ED_25519: &str = "ed25519";
+
+/// Represents hash of Iroha entities like `Block` or `Transaction.
+pub type Hash = [u8; HASH_LENGTH];
+
+pub enum KeyGenOption {
+    UseSeed(Vec<u8>),
+    FromPrivateKey(PrivateKey),
+}
+
+impl TryFrom<KeyGenOption> for UrsaKeyGenOption {
+    type Error = String;
+
+    fn try_from(key_gen_option: KeyGenOption) -> Result<Self, Self::Error> {
+        match key_gen_option {
+            KeyGenOption::UseSeed(seed) => Ok(UrsaKeyGenOption::UseSeed(seed)),
+            KeyGenOption::FromPrivateKey(key) => {
+                if key.digest_function == ED_25519 {
+                    Ok(UrsaKeyGenOption::FromSecretKey(UrsaPrivateKey(key.payload)))
+                } else {
+                    Err(format!(
+                        "Ursa does not support {} digest function.",
+                        key.digest_function
+                    ))
+                }
+            }
+        }
+    }
+}
+
+/// Pair of Public and Private keys.
+#[derive(Clone, Debug, Deserialize, Default)]
+pub struct KeyPair {
+    /// Public Key.
+    pub public_key: PublicKey,
+    /// Private Key.
+    pub private_key: PrivateKey,
+}
+
+impl KeyPair {
+    /// Generates a pair of Public and Private key.
+    /// Returns `Err(String)` with error message if failed.
+    pub fn generate(key_gen_option: Option<KeyGenOption>) -> Result<Self, String> {
```

Comment:
It is not a good practice to place `Option` inside function arguments. It means that no matter what variant is more common (None or Some in this case) - users will write something in any case and should know about this function arguments behavior depending on the value passed.

Better to split into two methods - one `generate()` and another one `generate_with_option(key_gen_option: KeyGenOption)`

Reply:
I decided to make it like that because it is a light wrapper around ursa generate, which uses this pattern of Option<KeyGenOption> - to make it similar. Do you still think it is worth dividing it into 2 functions?

Reply:
Yes, I insist. We can make our API better in terms of UX

Reply:
Ok, sure